Forms: Add notice after optimistic response status update

dhasilva opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Follow-up to #45912

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Fixes an issue where undoing an action reloaded the wrong page
* Addresses code review comments
* Adds a notice while updating the status, after optimistic update
* Sets the page to loading state if all items from the page are optimistically moved

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1763124260359199/1763124228.578279-slack-C052XEUUBL4

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to the seconds page of the inbox
* Trash a response or mark it as spam
* Check that a notice with a spinner icon is displayed until the action is finished on the server side
* Click on Undo
* Check that the second page items are kept (without the PR, the first page is loaded if spam/trash have only one page)
* Now select all items from the page and trash/mark as spam
* Check that the page displays the loading screen instead of the empty page

dhasilva committed:
optimizations from review

dhasilva committed:
add notice to pending action and update load state logic

dhasilva committed:
changelog

dhasilva committed:
fix js tests

dhasilva committed:
fix issue with wrong page loaded on undo

vianasw commented on projects/packages/forms/src/dashboard/inbox/stage/actions.tsx:
```diff
@@ -466,6 +567,10 @@ export const restoreAction: Action = {
 			invalidateCacheAndNavigate( registry, getCurrentQuery(), queryParams, 'trash' );
 		}
 
+		requestAnimationFrame( () => {
```
I think `requestAnimationFrame` doesn't guarantee cache invalidation is complete. It just schedules for the next paint. If cache invalidation takes longer, you might remove the pending action before data refetches, causing a brief flash of empty/old data.

dhasilva replied:
Yes, I had to use `getEntityRecords` to ensure the data is loaded before setting the pending action as resolved. Good catch. I was testing with the trash action, which reloaded faster.

dhasilva committed:
fix premature removal of pending actions

vianasw reviewed (changes requested):
Cache invalidation is a really hard problem! I tested it again, and the flickering still persists when I undo an action (mark as spam, send to trash). Upon investigation, here is what I discovered:
 1. Mark spam: API call → response arrives → receiveEntityRecords(spam) auto-called by WordPress
 2. Undo clicked: optimistic update → editEntityRecord(publish)
 3. First API's receiveEntityRecords(spam) arrives → overwrites the undo's optimistic edit → flicker!
 4. Undo API completes → receiveEntityRecords(publish)
 
I think I have a fix that involves waiting for the original action to complete, but adds a slight delay. I'll follow up on Slack.

vianasw committed:
Fix UI flicker when undoing status change actions

Fixes a race condition where clicking undo immediately after marking a
response as spam/trash would cause the UI to flicker. The item would
briefly disappear and reappear, and checkboxes would flash on all entries.

Root cause: When undo was clicked, the original action's API response
would arrive after the undo's optimistic update, overwriting it with
stale data. This created a sequence where the UI would update to the
wrong state (spam) before correcting itself (publish).

Solution: Make undo wait for the original action's refetch to complete
before starting. Implemented by:
- Tracking pending refetch promises in a Map by actionId
- Making undo onClick async and awaiting the original refetch
- Using undoTriggered flag to skip refetch if undo clicked quickly
- Removing original pending action state before starting undo

This ensures only one refetch result is applied, eliminating the flicker
while keeping undo responsive.
